smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/AbstractConfigGenerator.kt

```diff
@@ -0,0 +1,158 @@
+            .call { renderCompanionObject(writer) }
+            .call { renderBuilder(sortedProps, writer) }
+            .closeBlock("}")
+
+        writer.removeContext("configClass.name")
+    }
+
+    protected open fun renderCompanionObject(writer: KotlinWriter) {
+        writer.withBlock("public companion object {", "}") {
+            write("public inline operator fun invoke(block: Builder.() -> kotlin.Unit): #configClass.name:L = Builder().apply(block).build()")
+        }
+    }
+
+    /**
+     * register import statements from config properties
+     */
+    private fun addPropertyImports(props: Collection<ConfigProperty>, writer: KotlinWriter) {
+        props.forEach {
+            it.baseClass?.let { baseClass ->
+                writer.addImport(baseClass)
+            }
+            writer.addImport(it.symbol)
+            writer.addImportReferences(it.symbol, SymbolReference.ContextOption.USE)
+            it.additionalImports.forEach { symbol ->
+                writer.addImport(symbol)
+            }
+        }
+    }
+
+    protected open fun renderImmutableProperties(props: Collection<ConfigProperty>, writer: KotlinWriter) {
+        props.forEach { prop ->
+            val override = if (prop.requiresOverride) "override" else "public"
+
+            when (prop.propertyType) {
+                is ConfigPropertyType.SymbolDefault -> {
+                    writer.write("$override val #1L: #2P = builder.#1L", prop.propertyName, prop.symbol)
+                }
+                is ConfigPropertyType.ConstantValue -> {
+                    writer.write("$override val #1L: #2T = #3L", prop.propertyName, prop.symbol, prop.propertyType.value)
+                }
+                is ConfigPropertyType.Required -> {
+                    writer.write(
+                        "$override val #1L: #2T = requireNotNull(builder.#1L) { #3S }",
+                        prop.propertyName,
+                        prop.symbol,
+                        prop.propertyType.message ?: "${prop.propertyName} is a required configuration property",
+                    )
+                }
+                is ConfigPropertyType.RequiredWithDefault -> {
+                    writer.write(
+                        "$override val #1L: #2T = builder.#1L ?: #3L",
+                        prop.propertyName,
+                        prop.symbol,
+                        prop.propertyType.default,
+                    )
+                }
+                is ConfigPropertyType.Custom -> prop.propertyType.render(prop, writer)
+            }
+        }
+    }
+
+    protected open fun renderBuilder(props: Collection<ConfigProperty>, writer: KotlinWriter) {
+        val baseClasses = props
+            .mapNotNull { it.builderBaseClass?.name }
+            .sorted()
+            .toSet()
+            .joinToString(", ")
+
+        val formattedBaseClasses = if (baseClasses.isNotEmpty()) ": $baseClasses" else ""
```

Comment:
**Suggestion**: Duplicated code, consider extracting to helper method.

Reply:
It's nearly duplicated, uses different field from the property though. I'll take a look at cleaning it up.